untrusted/test_enclave/build.rs

```diff
@@ -0,0 +1,253 @@
+    match status.code().unwrap() {
+        0 => (),
+        _ => panic!("Failed to run edger8")
+    }
+
+    let basename = edl_file.as_ref().file_stem().unwrap().to_str().unwrap();
+    let trusted = out_dir.join(format!("{}_t.c", basename));
+    let untrusted = out_dir.join(format!("{}_u.c", basename));
+
+    EdgerFiles{trusted, untrusted}
+}
+
+/// Create enclave binary.  The binary is a shared library.
+///
+/// # Arguments
+///
+/// * `files` - The source files to include in the binary
+///
+/// # Returns
+/// The full path to resultant binary file.  This binary will be signed and
+/// ready for use in `sgx_create_enclave()`.
+fn build_enclave_binary<P>(files: P) -> PathBuf
+    where
+        P: IntoIterator,
+        P: Clone,
+        P::Item: AsRef<Path>, {
+
+    for file in files.clone() {
+        rerun_if_changed!(file.as_ref().as_os_str().to_str().expect("Invalid UTF-8 in enclave c file"));
+    }
+
+    // This `Build` builds a static library.  If we don't omit the
+    // `cargo_metadata` then this static library will be linked into
+    // the consuming crate. The enclave binary is meant to be a stand alone,
+    // so we do *not* want to link into the consuming crate.
+    // If one happens to link this in with the current crate, prepare for
+    // memory seg faults as the trusted (enclave) implementations will
+    // be directly linked in.
+    Build::new().files(files)
+        .include(format!("{}/include", sgx_library_path()))
+        .include(format!("{}/include/tlibc", sgx_library_path()))
+        .cargo_metadata(false) .shared_flag(true).compile("enclave");
+
+    let static_enclave = out_dir().join("libenclave.a");
+    let dynamic_enclave = build_dynamic_enclave_binary(static_enclave);
+    sign_enclave_binary(dynamic_enclave)
+}
+
+/// Create a dynamic version of the enclave.  This is *unsigned*.
+///
+/// See https://github.com/alexcrichton/cc-rs/issues/250 for lack of dynamic
+/// lib in cc crate
+///
+/// # Arguments
+///
+/// * `static_enclave` - The static enclave binary
+///
+/// # Returns
+/// The full path to resultant shared library file.
+fn build_dynamic_enclave_binary<P: AsRef<Path>>(static_enclave: P) -> PathBuf {
+    let mut dynamic_enclave = PathBuf::from(static_enclave.as_ref());
+    dynamic_enclave.set_extension("so");
+
+    let mut command = Command::new(ld_linker());
+    command
+        .arg("-o")
+        .arg(dynamic_enclave.to_str().expect("Invalid UTF-8 in static enclave path"))
+        .args(&["-z", "relro", "-z", "now", "-z", "noexecstack"])
+        .arg(&format!("-L{}/lib64/cve_2020_0551_load", sgx_library_path()))
+        .arg(&format!("-L{}/lib64", sgx_library_path()))
+        .arg("--no-undefined")
+        .arg("--nostdlib")
+        .arg("--start-group")
+        .args(&["--whole-archive", "-lsgx_trts_sim", "--no-whole-archive"])
+        .arg(static_enclave.as_ref().to_str().unwrap())
+        .args(&["-lsgx_tstdc", "-lsgx_tcxx", "-lsgx_tcrypto", "-lsgx_tservice_sim"])
```

Comment:
Do we always want the `_sim` libraries here?

Reply:
I have no problem opening the tests up to HW if we think it will be easy to utilize in CI.  If we end up not having a constant HW run going, I would be concerned that it breaks too easily.

I still have access to VM with HW sgx if we want to pursue that now

Reply:
It seems the GHA runners, both generic and self-hosted, support SGX. So if it's not too much extra work, supporting HW would be nice. It can also remain a TODO until we need it.

Reply:
We need to be able to run `cargo test` on desktops, which don't support SGX, so we need to keep the sim versions of these accessible.

Reply:
This is addressed now.  This required renaming the `sim` feature to `hw` and adding similar logic to the test_enclave so one can do `cargo test --features "hw" --features "test_enclave/hw"` to run the tests against the HW SDK